### PR TITLE
Subgraph Fusion

### PR DESCRIPTION
**Pull request by sancierra**
Pipeline that allows the fusion of not just two maps, but an arbitrary number of maps if fusible and with less restrictions than MapFusion. 

Functionality includes Expansion of Reduce Nodes into an inner reduce surrounded by an outer map consisting of the remaining axes. MultiExpansion then takes a set of maps and splits each of those into two nested maps with the outer map consisting of the maximum common map ranges and indices. Finally, SubgraphFusion fuses those outer maps together into one kernel, automatically creating necessary transients and connections to inputs and outputs.

Expansion of Reduce Nodes is implemented as a regular pattern_matching.Transformation, expansion and subgraph_fusion as subgraph transformations with their respective interfaces. pipeline.py contains an alternative quick method for accessing these transformations.

See subgraph_fusion.py for current implementation status.

### CODE REVIEW

**Comment by tbennun** [comment]
Great work! One note about the PR: we normally locate our tests under `tests/*/*_test.py`. In your case, I would use the existing transformations subfolder. This way, the tests are automatically run as part of our CI.

File: tests/transformations/sfusion_complex_test.py [hunk 1]
```diff
@@ -125,6 +125,7 @@ def test_quantitatively(sdfg, graph):
     csdfg = sdfg.compile()
     csdfg(A=A, B=B, C=C, OUT1 = OUT1, OUT2 = OUT2, OUT3 = OUT3, N=N, M=M, O=O)
 
+    sdfg.view()
```
Comment (tbennun) [inline tests/transformations/sfusion_complex_test.py:128]: :)

File: dace/transformation/subgraph/subgraph_fusion.py [hunk 2]
```diff
@@ -0,0 +1,902 @@
...
+                transient_to_transform = sdfg.data(data_name)
+                transient_to_transform.shape = new_data_shape
+                transient_to_transform.strides = new_data_strides
+                transient_to_transform.total_size = new_data_totalsize
+                transient_to_transform.offset = new_data_offset
+                transient_to_transform.lifetime = dtypes.AllocationLifetime.Scope
+                if schedule == dtypes.ScheduleType.GPU_Device:
+                    transient_to_transform.storage = self.cuda_transient_allocation
+                else:
+                    transient_to_transform.storage = self.cpu_transient_allocation
```
[the remark this answers is not in the document]
Comment (sancierra) [reply dace/transformation/subgraph/subgraph_fusion.py]: makes sense, fixed

File: dace/transformation/subgraph/subgraph_fusion.py [hunk 3]
```diff
@@ -0,0 +1,909 @@
...
+            into the map and their data dimension gets cropped.
+            Otherwise the original array is taken.
+
+            For every output respective connections are crated automatically.
+
+            :param sdfg: SDFG
+            :param graph: State
+            :param map_entries: Map Entries (class MapEntry) of the outer maps
+                                which we want to fuse
+            :param do_not_override: List of AccessNodes that are transient but
```
[the remark this answers is not in the document]
Comment (sancierra) [reply dace/transformation/subgraph/subgraph_fusion.py:451]: ok

**Review by tbennun — APPROVED** [review]

**Pull request merged**